src/flashdeberta/ops/flash_attention_varlen.py

```diff
@@ -164,42 +224,141 @@ def _fwd_kernel_deberta_disentangled_attention(
+                    BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 3, 4
+                else:
+                    BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 4, 4
             else:
-                if M <= 1024:
-                    BLOCK_M, BLOCK_N, num_stages, num_warps = 128, 32, 2, 4
+                if max_seqlen_q <= 1024:
+                    BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 2, 4
                 else:
-                    BLOCK_M, BLOCK_N, num_stages, num_warps = 128, 128, 3, 8
-    elif torch.cuda.get_device_capability() == (8, 6):
+                    BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 3, 8
+    elif capability[0] == 7:
         if not causal:
             if D <= 64:
-                BLOCK_M, BLOCK_N, num_stages, num_warps = 128, 64, 3, 4
+                BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 3, 4
             else:
-                BLOCK_M, BLOCK_N, num_stages, num_warps = 128, 32, 2, 4
-        else: # causal
+                BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 2, 4
+        else:  # causal
             if D <= 64:
                 BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 3, 4
             else:
-                BLOCK_M, BLOCK_N, num_stages, num_warps = 128, 32, 2, 4
+                BLOCK_M, BLOCK_N, num_stages, num_warps = 64, 64, 2, 4
     else:
-        BLOCK_M, BLOCK_N, num_stages, num_warps = 32, 32, 1, 4
+        BLOCK_M, BLOCK_N, num_stages, num_warps = 16, 16, 1, 4
+    
+    # Additional adjustments for variable-length sequences
+    
+    # For very sparse batches (many short sequences), reduce block size
+    avg_seq_len = total_tokens / max(1, torch.cuda.device_count())
+    if avg_seq_len < 256:
+        BLOCK_M = min(BLOCK_M, 64)
+        BLOCK_N = min(BLOCK_N, 32)
+        num_stages = max(1, num_stages - 1)  # Reduce stages to save memory
+    
+    # Calculate shared memory usage with current config
+    has_pos = disentangled
+    ATT_SPAN = att_span if has_pos else 0
+    
+    dtype = torch.float16  # Assuming float16 is used as in original code
+    
+    shared_mem_usage = calculate_shared_memory_usage_varlen(
+        BLOCK_M, BLOCK_N, D, num_stages, dtype,
+        has_c2p=has_pos, has_p2c=has_pos, ATT_SPAN=ATT_SPAN
+    )
+    
+    # If shared memory usage exceeds available, adjust parameters
+    # We prioritize reducing num_stages first, then block sizes
+    while shared_mem_usage > max_shared_memory and (BLOCK_M > 16 or BLOCK_N > 16 or num_stages > 1):
+        # First try reducing num_stages
+        if num_stages > 1:
+            num_stages -= 1
+        # Then try reducing block sizes
+        elif BLOCK_M > 32 and BLOCK_N > 32:
+            BLOCK_M //= 2
+            BLOCK_N //= 2
+        elif BLOCK_M > 32:
+            BLOCK_M //= 2
+        elif BLOCK_N > 32:
+            BLOCK_N //= 2
+        elif BLOCK_M > 16:
+            BLOCK_M //= 2
+        elif BLOCK_N > 16:
+            BLOCK_N //= 2
```

Comment:
```suggestion
        else:
            BLOCK_M //= 2
            BLOCK_N //= 2
```

As blocks need to remain square and you already have an exit condition so your min block size is 8x8